straxbra/utils.py

```diff
@@ -8,10 +8,22 @@
 
 __client = MongoClient(os.environ['MONGO_DAQ_URI'])
 db = __client['xebra_daq']
-experiment = 'xebra'
+experiments = {
+        0: {'name': 'xebra', 'n_pmts': 8, 'drift_length': 7}, # drift_lengths in cm
+        1: {'name':  'htpc', 'n_pmts': 2, 'drift_length': np.nan}  # TODO: insert correct drift_length
+        }
+
 MAX_RUN_ID = 999999  # because reasons
 
+def _GetExperiment(run_id, info):
```

Comment:
How is this more efficient than just changing `straxbra.utils.experiment`? This hard-caps XeBRA's lifespan, and you'll need to change it again once single-phase is a thing. Also this might cause havoc in the dispatcher.

Reply:
I don't get what you're suggesting. This variable doesn't seem to be called by strax-contexts when calling these functions, is it? That's my main problem.

What I did is to change `experiment = 'xebra'` to the dictionary that gets the right data (that was hard coded before for xebra) for each experiment based on the run_id. I see your point that it hard-caps the lifespan of xebra (some day), but I wouldn't know how to pass it the experiments parameters without this run-dependency. When you look into how strax (config.py, get_dafault, line 106) calls these functions they are only called with the run_id and it seems not to be provided that one passes something else.

Ok the`_GetExperiment` function might not be necessary, but to me it looks way nicer than this nested
`experiments[int(str(run_id)[0])]['experiment']` since it's used in a few places.

Was any of this what you ment?

Reply:
The context is free to overwrite `straxbra.utils.experiment` when it's created, this is how the 1T and nT contexts account for differing numbers of PMTs (for instance). Having the experiment encoded into the run id is going to make life difficult for the dispatcher because it makes it much more difficult to assign run ids to new runs.

Reply:
Understood. Comes in next commit :+1:

Reply:
This doesn't work. Utils is not part of straxbra (not exported). This will thus fail:
```
class HtpcContext(strax.Context):
    ...

    def __init__(self, *args, *kwargs):
        self.utils.experiment = 'htpc'
        ....
```
(same obviously for `straxbra.utils.experiment`)
I've been looking how e.g. `GetGains` is actually called when used by a plugin like this:

```
from . import utils

...
@export
@strax.takes_config(
        strax.Option(.....),
        strax.Option('to_pe', track=False,
                     default_by_run=utils.GetGains,
                     help='PMT gains'),
)
class Peaks(strax.Plugin):
```

That's done in strax.config.get_default:
```
def get_default(self, run_id=None):
        """Return default value for the option"""

        ...

        if self.default_by_run is not OMITTED:
            if callable(self.default_by_run):
                return self.default_by_run(run_id)
```

- How about giving it self.config there and use whatever config we need in the utils function?
- Another option is probably to let the context also inherit from utils and set the option in the context as you suggested. I don't know if this works though because the plugins are registered and called by strax, but not a parent class of the context.

Reply:
- `utils` is not an attribute of the context, so `self.utils.experiment='htpc'` isn't valid
- `utils` also isn't a class, so the context can't inherit from it. Maybe you mean 'import'.

Reply:
Done. Seems to work.